tt_metal/graph_tracking.hpp

```diff
@@ -0,0 +1,136 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <string>
+#include <any>
+#include <span>
+#include <string_view>
+
+#include "tt_metal/common/core_coord.h"
+#include "tt_metal/impl/buffers/buffer.hpp"
+
+namespace tt::tt_metal {
+
+    class Program;
+    class IGraphProcessor{
+    public:
+        enum class RunMode {
+            REAL,
+            FAKE
+        };
+
+        IGraphProcessor() = default;
+
+        virtual void track_allocate(tt::tt_metal::Buffer* buffer, bool bottom_up) {};
+
+        virtual void track_deallocate(tt::tt_metal::Buffer* buffer) {};
+
+        virtual void track_allocate_cb(const CoreRangeSet &core_range_set, uint64_t addr, uint64_t size) {};
+
+        virtual void track_deallocate_cb() {};
+
+        virtual void track_program(tt::tt_metal::Program* program) {};
+
+        virtual void track_begin_op(std::string_view function_name, std::span<std::any> input_parameters) {};
```

Comment:
why not keep it in as parameter pack like you did below and then convert it in the impl:
`void GraphProcessor::track_begin_op(std::string_view function_name, std::span<std::any> input_parameters)`

Reply:
It should be a virtual function.